projection/core/src/main/scala/com/lightbend/lagom/internal/projection/ProjectionRegistryActor.scala

```diff
@@ -4,128 +4,186 @@
+  // TODO: make timeout configurable (5.second is too aggressive in big clusters)
+  val writeMajority: WriteConsistency  = WriteMajority(timeout = 5.seconds)
+  val readConsistency: ReadConsistency = ReadLocal
+
+  // (a) Replicator contains data of all workers
+  private val RequestedStatusDataKey: LWWMapKey[WorkerKey, Status] =
+    LWWMapKey[WorkerKey, Status]("projection-registry-desired-status")
+  private val ObservedStatusDataKey: LWWMapKey[WorkerKey, Status] =
+    LWWMapKey[WorkerKey, Status]("projection-registry-observed-status")
+  private val NameIndexDataKey: LWWMapKey[WorkerKey, WorkerCoordinates] =
+    LWWMapKey[WorkerKey, WorkerCoordinates]("projection-registry-name-index")
+  replicator ! Subscribe(RequestedStatusDataKey, self)
+  replicator ! Subscribe(ObservedStatusDataKey, self)
+  replicator ! Subscribe(NameIndexDataKey, self)
+
+  // (b) Keep a local copy to simplify the implementation of some ops
+  var requestedStatusLocalCopy: Map[WorkerKey, Status] = Map.empty[WorkerKey, Status]
+  var observedStatusLocalCopy: Map[WorkerKey, Status]  = Map.empty[WorkerKey, Status]
+
+  var nameIndexLocalCopy: Map[WorkerKey, WorkerCoordinates] = Map.empty[WorkerKey, WorkerCoordinates]
+
+  // (c) Actor indices contain only data of workers running locally
+  var actorIndex: Map[WorkerKey, ActorRef] = Map.empty[WorkerKey, ActorRef]
   // required to handle Terminate(deadActor)
-  var actorReverseIndex: Map[ActorRef, WorkerMetadata] = Map.empty[ActorRef, WorkerMetadata]
+  var reversedActorIndex: Map[ActorRef, WorkerKey] = Map.empty[ActorRef, WorkerKey]
 
   override def receive: Receive = {
-    case RegisterProjection(streamName, projectionName, workerName) =>
-      val metadata = WorkerMetadata(streamName, projectionName, workerName)
-      // when registering a projection worker, we default to state==enabled
-      val writeMajority = WriteMajority(timeout = 5.seconds)
-      replicator ! Update(DataKey, PNCounterMap.empty[WorkerMetadata], writeMajority)(
-        //TODO: read the default state from a desired _initial state_
-        // https://github.com/lagom/lagom/issues/1744
-        _.increment(node, metadata, 1)
-      )
-      // keep track and watch
-      actorIndex = actorIndex.updated(metadata, sender)
-      actorReverseIndex = actorReverseIndex.updated(sender, metadata)
+
+    case RegisterProjectionWorker(coordinates) =>
+      log.debug(s"Registering worker $coordinates to [${sender().path.toString}]")
+      // keep track
+      val workerKey: WorkerKey = coordinates.asKey
+      updateLWWMapForNameIndex(workerKey, coordinates)
+      actorIndex = actorIndex.updated(workerKey, sender())
+      reversedActorIndex = reversedActorIndex.updated(sender, workerKey)
+      // when worker registers, we must reply with the requested status (if it's been set already, or Started if not).
+      // TODO: parameterize the default initial status
+      val requestedStatus = requestedStatusLocalCopy.getOrElse(workerKey, Started)
```

Comment:
I don't see how we could have a WokerHolder registering itself and at the same time be already registered in the CRDT.

Reply:
so I'm trying to think about which situation this could happen. 

I see two scenarios:

### Node crashes

When that happens, the worker holder and registry and everything just evaporates in front of our eyes. 
At some point, the worker will be instantiated on another node and it will upon registration learn about its requested state.

### Gracefull shutdown

In case of a split-brain, a node will be gracefully removed and the WorkHolder will be terminated. Then the Terminated message will be handled, on line 157. This will cause the CRDT to be updated with the requested state to Stopped. Normally, the data won't be propagated because of split-brain. But we may have the situation in which the split is detected, the shutdown starts and the network recovers and the data gets replicated. In such a scenario, the worker starts on another node, but learns that it should be in 'Stopped' state. 

I'm not sure how realistic is this. Does SBR cuts all communication when it decides that there is a split? 

### Down scale

Down scale is the more problematic. We gracefully shutdown a node, the WorkHolder is terminated, its requested state moves to Stopped. And it re-initializes as Stopped on another node.

Reply:
> I don't see how we could have a WokerHolder registering itself and at the same time be already registered in the CRDT.

Imagine this case:

1. the local `ProjectionRegistry` is up to date with all `ddata`in the cluster. 
2. another node dies
3. the shards on that node are relocated and one shard ends on our local node

In that case, where a long-running node adopts a shard that relocated from a dying node we have a WokerHolder registering itself and we already have it's requested status tracked in `requestedStatusLocalCopy`.

Reply:
> We gracefully shut down a node, the WorkHolder is terminated, its requested state moves to Stopped. And it re-initializes as Stopped on another node. This is certainly not what we want.

This is not correct.

When a worker dies, its requested status doesn't change. Its _observed_ state, OTOH, does change. That's the key difference between the _requested_ and the _observed_: what I _want_ VS what I _have.

Reply:
@ignasi35, I see now your point. We update the observed state, not the requested. The code as is is therefore correct. 

We still may miss the observed state message on a downscale. The WorkerHolder is not a child of the registry and therefore it may terminate first.